src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+
+        # Need to sort lists by highest_position_nodes value, and then
+        # by lowest value if tied.
+        # Based upon
+        # https://stackoverflow.com/questions/9764298/how-to-sort-two-
+        # lists-which-reference-each-other-in-the-exact-same-way
+        if len(dependent_nodes) > 0:
+            sorted_highest_positions, sorted_lowest_positions, \
+                    sorted_dependency_pairs = (list(t) for t in
+                                               zip(*sorted(zip(
+                                                   highest_position_nodes,
+                                                   lowest_position_nodes,
+                                                   dependent_nodes))))
+        else:
+            # We have no invalid dependencies so we can return early
+            return
+
+        taskwait_location_nodes = []
+        taskwait_location_abs_pos = []
+        for taskwait in self.walk(OMPTaskwaitDirective):
+            taskwait_location_nodes.append(taskwait)
+            taskwait_location_abs_pos.append(taskwait.abs_position)
+        # Add the first node to have a taskwait placed in front of it into the
+        # list
+        taskwait_location_nodes.append(sorted_dependency_pairs[0][1])
+        taskwait_location_abs_pos.append(sorted_highest_positions[0])
+        for index, pairs in enumerate(sorted_dependency_pairs[1:]):
+            # Add 1 to index here because we're looking from [1:]
+            lo_abs_pos = sorted_lowest_positions[index+1]
+            hi_abs_pos = sorted_highest_positions[index+1]
+            satisfied = False
+            for ind, taskwait_loc in enumerate(taskwait_location_nodes):
+                if (taskwait_location_abs_pos[ind] <= hi_abs_pos and
+                        taskwait_location_abs_pos[ind] >= lo_abs_pos):
+                    # We have a taskwait meant to be placed here that is
+                    # potentially already satisfied. To check we need to
+                    # ensure that the ancestor schedules of the nodes
+                    # are identical
+                    if (pairs[0].ancestor(Schedule) is
+                            taskwait_loc.ancestor(Schedule)):
+                        satisfied = True
+                        break
+            # If we didn't find a taskwait we plan to add that satisfies this
+            # dependency, add it to the list
+            if not satisfied:
+                taskwait_location_nodes.append(pairs[1])
+                taskwait_location_abs_pos.append(hi_abs_pos)
+        # Now loop through the list in reverse and add taskwaits
+        taskwait_location_nodes.reverse()
+        for taskwait_loc in taskwait_location_nodes:
+            node_parent = taskwait_loc.parent
+            loc = taskwait_loc.position
+            node_parent.addchild(OMPTaskwaitDirective(), loc)
+            # As the node_parent tree has already been lowered and we're adding
```

Comment:
Is that right? I would have thought that `OMPTaskwaitDirective` was already a language-level node and doesn't need lowering. Is there something I've missed?

Reply:
Ah, this might just be my misunderstanding - it is a leaf node so I think I can just remove the lowering call.